wallet/controllers_v3_test.go

```diff
@@ -260,3 +260,77 @@ func TestGetWalletV3(t *testing.T) {
 		must(t, "invalid response", fmt.Errorf("expected 201, got %d", resp.StatusCode))
 	}
 }
+
+func TestLinkBitFlierWalletV3(t *testing.T) {
+	mockCtrl := gomock.NewController(t)
+	defer mockCtrl.Finish()
+	var (
+		secret      = []byte("a jwt secret")
+		tokenString = "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJhY2NvdW50X2hhc2giOiIxMjM0NTY3ODkwIiwiZGVwb3NpdF9pZCI6ImMzZmQ5YTBhLWM0ZDItNDdmYy1iZjdlLTlhZTJiYzk3MzNhMSIsImV4dGVybmFsX2FjY291bnRfaWQiOiIyIiwicmVxdWVzdF9pZCI6IjEiLCJ0aW1lc3RhbXAiOiIxOTcwLTAxLTAxVDAwOjAwOjAwWiJ9.1-PTPLYhMKJkscgoQAY8gQic0cPBIceD4eVTcJ9J-0k"
+		db, mock, _ = sqlmock.New()
+		datastore   = wallet.Datastore(
+			&wallet.Postgres{
+				grantserver.Postgres{
+					DB: sqlx.NewDb(db, "postgres"),
+				},
+			})
+		roDatastore = wallet.ReadOnlyDatastore(
+			&wallet.Postgres{
+				grantserver.Postgres{
+					DB: sqlx.NewDb(db, "postgres"),
+				},
+			})
+
+		// add the datastore to the context
+		ctx = context.WithValue(context.Background(), appctx.BitFlierJWTKeyCTXKey, string(secret))
+		r   = httptest.NewRequest(
+			"POST",
+			"/v3/wallet/bitflier/7def9cda-6a14-4fa1-be86-43da80e56d2c/claim",
+			bytes.NewBufferString(fmt.Sprintf(`
+				{
+					"linkingInfo": "%s"
+				}`, tokenString)),
+		)
+		mockReputation = mockreputation.NewMockClient(mockCtrl)
+		handler        = wallet.LinkBitFlierDepositAccountV3(&wallet.Service{
+			Datastore: datastore,
+		})
+		w = httptest.NewRecorder()
+
+		walletClaimNamespace = uuid.Must(uuid.FromString("c39b298b-b625-42e9-a463-69c7726e5ddc"))
+		idTo, _              = uuid.FromString("c3fd9a0a-c4d2-47fc-bf7e-9ae2bc9733a1")
+		providerLinkingID    = uuid.NewV5(walletClaimNamespace, "1234567890")
+		idFrom, _            = uuid.FromString("7def9cda-6a14-4fa1-be86-43da80e56d2c")
+		rows                 = sqlmock.NewRows([]string{"id", "provider", "provider_id", "public_key", "provider_linking_id", "anonymous_address"}).
+					AddRow(idFrom, "bitflier", "", "12345", providerLinkingID, idTo)
+	)
+
+	mock.ExpectQuery("^select (.+)").WithArgs(idFrom).WillReturnRows(rows)
+
+	mock.ExpectBegin()
+	mock.ExpectQuery("^select (.+)").WithArgs(providerLinkingID).WillReturnRows(rows)
+
+	// txHasDestination
+	var hasDestRows = sqlmock.NewRows([]string{"bool"}).AddRow(true)
+	mock.ExpectQuery("^select (.+)").WithArgs(idFrom).WillReturnRows(hasDestRows)
+
+	mock.ExpectExec("^UPDATE (.+)").WithArgs(idFrom, providerLinkingID, "bitflier").WillReturnResult(sqlmock.NewResult(1, 1))
+	mock.ExpectCommit()
+
+	ctx = context.WithValue(ctx, appctx.DatastoreCTXKey, datastore)
+	ctx = context.WithValue(ctx, appctx.RODatastoreCTXKey, roDatastore)
+	ctx = context.WithValue(ctx, appctx.ReputationClientCTXKey, mockReputation)
```

Comment:
I personally like the use of mocks primarily so that I don't need to stand up an entire environment to run these tests.  I also like the ability to inject errors into responses to test error conditions.  In these particular tests (wallet v3) there is no datastore actually setup, and these do not have the integration build flag.

Reply:
I suggest we chat about the pros and cons at some point of both methods.  I feel unit tests are better suited for mocking, and we could cut down on our test/build times as well as get more detailed edge case testing if we go with a mocking framework.  Integration tests are still required, but for happy path full flow integration tests imo, not just testing datastore model to postgres integrations.  Regardless, the v3 wallet testing as a whole is using mocking for integration tests, so I suggest we make a plan outside this PR to address if we want to change this.